3p-ecosystem/src/main/java/com/google/homesampleapp/screens/device/DeviceFragment.kt

```diff
@@ -332,68 +335,206 @@ class DeviceFragment : Fragment() {
+    val lastUpdatedDeviceState by devicesStateRepository.lastUpdatedDeviceState.observeAsState()
+    val shareDeviceStatus by viewModel.shareDeviceStatus.observeAsState()
+
+    binding.topAppBar.title = selectedDevice?.device?.name
 
-  private fun updateDeviceInfo(deviceState: DeviceState?) {
+    DeviceScreen(selectedDeviceId, selectedDevice, lastUpdatedDeviceState, shareDeviceStatus)
+  }
+
+  @Composable
+  private fun DeviceScreen(
+    selectedDeviceId: Long?,
+    deviceUiModel: DeviceUiModel?,
+    deviceState: DeviceState?,
+    shareDeviceStatus: TaskStatus?
+  ) {
     // The DeviceUiModel is not updated whenever we observe changes in the state of the device.
     // This is an issue for the "Inspect Device" onClick listener which relies on the device
     // state to decide whether to show a dialog stating that the device is offline and therefore
     // the inspect screen cannot be shown, or go show the inspect information (when device is
     // online).
     // This is why the state of the device is cached in local variables.
-    if (selectedDeviceViewModel.selectedDeviceIdLiveData.value == -1L) {
+    if (selectedDeviceId == null || selectedDeviceId == -1L) {
       // Device was just removed, nothing to do. We'll move to HomeFragment.
       isOnline = false
       return
     }
-    val deviceUiModel = selectedDeviceViewModel.selectedDeviceLiveData.value
 
     // Device state
     deviceUiModel?.let {
       isOnline =
-          when (deviceState) {
-            null -> deviceUiModel.isOnline
-            else -> deviceState.online
-          }
+        when (deviceState) {
+          null -> deviceUiModel.isOnline
+          else -> deviceState.online
+        }
       isOn =
-          when (deviceState) {
-            null -> deviceUiModel.isOn
-            else -> deviceState.on
-          }
+        when (deviceState) {
+          null -> deviceUiModel.isOn
+          else -> deviceState.on
+        }
 
-      binding.topAppBar.title = deviceUiModel.device.name
+      Timber.d("DeviceScreen [${deviceUiModel.toString()}]")
+      Column {
+        OnOffStateSection(deviceUiModel, deviceState, {
+          viewModel.updateDeviceStateOn(deviceUiModel, it)
+        })
+        ShareSection(id = deviceUiModel.device.deviceId, name = deviceUiModel.device.name, shareDeviceStatus)
+        // TODO: Use HorizontalDivider when it becomes part of the stable Compose BOM.
+        Spacer(
+          modifier = Modifier,
+        )
+        TechnicalInfoSection(deviceUiModel.device)
+        RemoveDeviceSection({ removeDevice() })
+      }
+    }
+  }
 
-      val shapeOffDrawable = getDrawable("device_item_shape_off")
-      val shapeOnDrawable = getDrawable("device_item_shape_on")
-      val shapeDrawable = if (isOnline && isOn) shapeOnDrawable else shapeOffDrawable
+  @Composable
+  private fun OnOffStateSection(
+    deviceUiModel: DeviceUiModel?,
```

Comment:
Following the suggestion to move the `isOnline` and `isOn` fields into composition above, this signature of this method can become:

```
@Composable
private fun OnOffStateSection(
     isOnline: Boolean,
     isOn: Boolean,
     onStateChange: ((Boolean) -> Unit)?
 )   
```

Reply:
Done.